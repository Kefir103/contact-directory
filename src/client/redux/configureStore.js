import { createStore, applyMiddleware, compose } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension/index';
import { rootReducer } from './reducers/rootReducer';
import thunk from 'redux-thunk';

let store;

export default function configureStore(initialState) {
    if (store) {
        return store;
    }

    store =
        process.env.NODE_ENV === 'production'
            ? createStore(rootReducer, initialState, compose(applyMiddleware(thunk)))
            : createStore(rootReducer, initialState, composeWithDevTools(applyMiddleware(thunk)));

    return store;
}
