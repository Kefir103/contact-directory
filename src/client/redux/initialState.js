export const initialState = {
    filter: {
        sortingMap: new Map([]),
        filterText: '',
        filterFields: [],
        currentPage: 1,
    },
    data: {
        currentElement: {
            elementInfo: undefined,
            elementIndex: -1,
        },
        elements: [],
        currentElements: [],
        filteredElements: undefined,
        inputContainer: {
            isAddButtonDisabled: false,
            isFormOpen: false,
            inputElement: {
                id: 0,
                firstName: '',
                lastName: '',
                email: '',
                phone: '',
            },
            validInputs: {
                id: false,
                firstName: false,
                lastName: false,
                email: false,
                phone: false,
            },
            addingStatus: {
                addingStatusText: '',
                isError: false,
            },
        },
    },
    appStatus: {
        isLoading: false,
        error: undefined,
    },
};
