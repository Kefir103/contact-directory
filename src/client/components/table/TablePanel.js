import React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { setSortingMap } from '../../redux/actions/filterActions';
import {setAppElements, setPageCount} from '../../redux/actions/dataActions';
import {getAppElements, getPagesCount} from '../../functions/dataFunctions';
import {catchError} from "../../redux/actions/appStatusActions";

const TablePanel = (props) => {
    const tableButtonClickHandle = (event) => {
        event.preventDefault();
        try {
            const newMap = new Map(props.sortingMap);

            if (!newMap.has(event.target.value)) {
                newMap.set(event.target.value, 'asc');
            } else if (newMap.get(event.target.value) === 'asc') {
                newMap.set(event.target.value, 'desc');
            } else if (newMap.get(event.target.value) === 'desc') {
                newMap.delete(event.target.value);
            }
            props.actions.setSortingMap(newMap.entries());

            const newArrayOfAppElements = props.filterText
                ? getAppElements(
                    props.fullData,
                    { filterText: props.filterText, filterFields: props.filterFields },
                    newMap
                )
                : getAppElements(props.fullData, null, newMap);
            const pageCount = getPagesCount(newArrayOfAppElements);
            props.actions.setAppElements(newArrayOfAppElements);
            props.actions.setPageCount(pageCount);
        } catch (error) {
            props.actions.catchError(error);
        }
    };

    const getArrow = (field) => {
        if (!props.sortingMap.has(field)) {
            return '';
        } else if (props.sortingMap.get(field) === 'asc') {
            return 'up-arrow';
        }
        return 'down-arrow';
    };

    return (
        <div className={'table-row panel'}>
            <button
                className={`table-cell-button id ${getArrow('id')}`}
                value={'id'}
                onClick={tableButtonClickHandle}>
                Id
            </button>
            <button
                className={`table-cell-button ${getArrow('firstName')}`}
                value={'firstName'}
                onClick={tableButtonClickHandle}>
                First Name
            </button>
            <button
                className={`table-cell-button ${getArrow('lastName')}`}
                value={'lastName'}
                onClick={tableButtonClickHandle}>
                Last Name
            </button>
            <button
                className={`table-cell-button ${getArrow('email')}`}
                value={'email'}
                onClick={tableButtonClickHandle}>
                Email
            </button>
            <button
                className={`table-cell-button ${getArrow('phone')}`}
                value={'phone'}
                onClick={tableButtonClickHandle}>
                Phone
            </button>
        </div>
    );
};

const mapStateToProps = (state) => {
    return {
        sortingMap: state.filter.sortingMap,
        filterText: state.filter.filterText,
        filterFields: state.filter.filterFields,
        fullData: state.data.fullData,
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        actions: bindActionCreators(
            {
                setSortingMap,
                setAppElements,
                setPageCount,
                catchError,
            },
            dispatch
        ),
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(TablePanel);
