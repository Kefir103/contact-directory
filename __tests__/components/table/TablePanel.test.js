import React from 'react';
import { shallow, render } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import TablePanel from '../../../src/client/components/table/TablePanel';
import { initialState } from '../../../src/client/redux/initialState';

const mockStore = configureMockStore();

describe('<TablePanel />', () => {
    test('should render correctly without selected sortingMap', () => {
        const store = mockStore({ ...initialState });
        const wrapper = shallow(<TablePanel store={store} />);
        const component = render(wrapper);
        expect(component).toMatchSnapshot();
    });
    test('should render correctly with selected sortingMap with asc and desc values', () => {
        const store = mockStore({
            ...initialState,
            filter: {
                ...initialState.filter,
                sortingMap: new Map([
                    ['id', 'asc'],
                    ['firstName', 'asc'],
                    ['lastName', 'asc'],
                    ['email', 'desc'],
                    ['phone', 'desc'],
                ]),
            },
        });
        const wrapper = shallow(<TablePanel store={store} />);
        const component = render(wrapper);
        expect(component).toMatchSnapshot();
    });
});
