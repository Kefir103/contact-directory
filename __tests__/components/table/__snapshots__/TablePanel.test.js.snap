// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TablePanel /> should render correctly with selected sortingMap with asc and desc values 1`] = `
<div
  class="table-row panel"
>
  <button
    class="table-cell-button id up-arrow"
    value="id"
  >
    Id
  </button>
  <button
    class="table-cell-button up-arrow"
    value="firstName"
  >
    First Name
  </button>
  <button
    class="table-cell-button up-arrow"
    value="lastName"
  >
    Last Name
  </button>
  <button
    class="table-cell-button down-arrow"
    value="email"
  >
    Email
  </button>
  <button
    class="table-cell-button down-arrow"
    value="phone"
  >
    Phone
  </button>
</div>
`;

exports[`<TablePanel /> should render correctly without selected sortingMap 1`] = `
<div
  class="table-row panel"
>
  <button
    class="table-cell-button id "
    value="id"
  >
    Id
  </button>
  <button
    class="table-cell-button "
    value="firstName"
  >
    First Name
  </button>
  <button
    class="table-cell-button "
    value="lastName"
  >
    Last Name
  </button>
  <button
    class="table-cell-button "
    value="email"
  >
    Email
  </button>
  <button
    class="table-cell-button "
    value="phone"
  >
    Phone
  </button>
</div>
`;
